### PR TITLE
Prometheus: Rework interval and step computation

### PR DESCRIPTION
**Pull request by alin-amana**
Rework the interaction between auto interval (computed based on graph resolution), min interval (where specified, per query) and intervalFactor (AKA resolution, where specified, per query).

As a bonus, have `$__interval` and `$__interval_ms` reflect the actual interval (not only the auto interval), taking into account min interval and Prometheus' 11k data points limit.

I was mostly concerned with the latter issue, namely ensuring there exists a reliable way of covering all data points in the range exactly once. (I'll explain this in more detail below.) But then I noticed that the way the interval and step were determined was rather inconsistent. More specifically, I noticed a couple of issues with the current implementation:

1. The stated goal of the resolution dropdown (AKA `intervalFactor`) is to reduce the load on Prometheus by requesting 1 data point for every X pixels (more specifically, a 1 in X sample). However, when the auto interval is 10+ seconds an 1/10 resolution will be ignored and as many data points will be requested as can be represented on the graph. This will usually happen when the range is longer than a few hours, which is exactly when one would most likely want to reduce Prometheus load.
1. There is weird behavior when the 11k data points limit comes into play, with the resolution being ignored again (and the actual step parameter not even being a multiple of the resolution, which might be an expectation if one wants to match Prometheus' scraping/eval intervals).

Now getting back to having a reliable way of covering all data points in the range exactly once, I have a recorded rule that quite reliably computes the increases in histogram buckets. If I represent that directly on a bar graph, I can get an almost exact 1:1 representation of how many requests have been received in every 10 second bucket. Now I want to have that reliably aggregated over arbitrary intervals that are multiples of 10 seconds. Problem is there is no reliable way of doing a `sum_over_time()` over the exact step used in the graph: `$__interval` doesn't take into account "Min step"; and a separately defined template variable won't be able to predict the auto interval. So I'm either getting overlapping or non-adjacent intervals. With `$__interval` matching the exact step used in the Prometheus query, I can now do `sum_over_time(my_metric[$__interval])` and be sure I'm not going to be under- or overcounting.

### CODE REVIEW

**Comment by alin-amana** [comment]
Friendly ping.

**Comment by torkelo** [comment]
hm.. maybe we should move resolution option from query to panel level  (like min interval & max data points for other datasources), that way it can be used to calculate the interval that is sent to the data source.

**Comment by bergquist** [comment]
ping @RichiH @juliusv @brian-brazil @SuperQ any opinions on this?

**Comment by alin-amana** [comment]
It's your call, I guess. :o)

But as quoted above the point of the resolution option is to reduce load on Prometheus for heavy queries. It is quite possible (and most likely) that the majority of panels on a dashboard are pretty lightweight and there are one or two that are really slow. In that case one might still want full resolution for the light queries and lower resolution for the heavy ones.

**Comment by alin-amana** [comment]
Prometheus people? Any input?

**Comment by RichiH** [comment]
I didn't test the code, but the functional description is sane.

**Comment by torkelo** [comment]
I like this but it's a bit tricky as to what implications this have. The $__interval and $__interval_ms is something the panel calculates and send to the data source in the query call, and if the data source modifies this there is no way for the panel to know. 

Kind of would like to introduce a resolution property on the panel level that could be used when calculating $__interval and $__interval_ms , but that would be a breaking change for all prometheus queries that use resolution. Could be a migration code that moves the highest resolution from a query to the panel. 

The other thing to consider is how this affects alerting, there is interval variables calculated in the backend for alerting requests as well (that also depend on alert query time range). We need the same logic there as well to be consistent.

**Comment by alin-amana** [comment]
I understand your concerns, well at least some of them. (I have no experience with Grafana alerts, so I don't have much to comment on that.)

Regarding the implications of the data source modifying the query interval, this is already happening, unfortunately. Currently it's only the sample resolution that is adjusted by the Prometheus data source, not the query itself. (It is affected by the per-query resolution and by Prometheus' 11k samples limit.) But, since this is already happening anyway, I am arguing that it's better to be able to also update the query to match the sampling. If you don't want that for some reason you can use either a fixed value in your query or one determined by a template variable, independent of the panel computed interval.

Or, if you think there is benefit in having a consistent interval across all queries for a panel, specifically controlled by the panel, then we could have 2 different $__interval auto variables: the panel-wide auto interval and the query specific interval, for Prometheus only (or maybe not, I'm only using Grafana with Prometheus, so far).

Moving on to your second point, I strongly believe there is good reason to have the option to adjust resolution per query. When using multiple queries in a panel, some of the time they will be quite similar (e.g. different quantiles on a latency graph) but some of the time they will be quite different, both in terms of the query itself and the load it puts on Prometheus. In the latter case, it is useful to be able to compromise on the resolution of some queries but not all.

Also, I'm not sure I explained this well enough in my initial comment: $__interval and $__interval_ms are not affected by the resolution in any way. (Actually they are currently, but that is a bug I'm trying to fix with this PR.) So I don't entirely understand the reasoning for moving the resolution up to the panel level.

To summarize my opinion (because that's all it is), the interval should really only be determined by the panel plus the min interval configured at query level (and the 11k sample limit, but that rarely applies). The resolution is very useful to have at a query level to reduce the load on Prometheus and it shouldn't affect the interval in any way (except tweaking the way the 11k limit is applied). And alerts I don't know about, but if I were to guess, all they would have in common with this logic is the 11k sample limit.

Apologies for the rant.

**Comment by torkelo** [comment]
> Also, I'm not sure I explained this well enough in my initial comment: $__interval and $__interval_ms are not affected by the resolution in any way. (Actually they are currently, but that is a bug I'm trying to fix with this PR.) So I don't entirely understand the reasoning for moving the resolution up to the panel level.

The reasoning is that these variables are "panel scoped", and could work in panel titles and other panel options (they don't right now thought but could). So would be confusing if they could be modified per query and become out of sync with that the panel generated. 

> The resolution is very useful to have at a query level to reduce the load on Prometheus and it shouldn't affect the interval in any way (except tweaking the way the 11k limit is applied). 

It shoudn't? I thought that was the main point of this PR to have resolution changes keep interval and inteval_ms in sync with the new step value. Might be confused or misunderstood something. 

> And alerts I don't know about, but if I were to guess, all they would have in common with this logic is the 11k sample limit.

The code that issues Prometheus queries in the Go backend should behave the same as the frontend code. So if prometheus data source code modifies the build in $__interval and $__interval_ms variables this needs to be done in the backend as well.

**Comment by alin-amana** [comment]
> > Also, I'm not sure I explained this well enough in my initial comment: $__interval and $__interval_ms are not affected by the resolution in any way. (Actually they are currently, but that is a bug I'm trying to fix with this PR.) So I don't entirely understand the reasoning for moving the resolution up to the panel level.
>
> The reasoning is that these variables are "panel scoped", and could work in panel titles and other panel options (they don't right now thought but could). So would be confusing if they could be modified per query and become out of sync with that the panel generated.

OK, I do understand now. I did in fact run into this myself (where I would have liked to explicitly display the interval in the title or elsewhere).

> > The resolution is very useful to have at a query level to reduce the load on Prometheus and it shouldn't affect the interval in any way (except tweaking the way the 11k limit is applied).
>
> It shoudn't? I thought that was the main point of this PR to have resolution changes keep interval and inteval_ms in sync with the new step value. Might be confused or misunderstood something.

It seems like a misunderstanding, yes. If you look at the code, `adjustedInterval` is computed in `adjustInterval()` and there are 2 things that affect it: one is `minInterval` ("Min step" in the UI) and the other is the 11k sample limit. The latter is Prometheus specific and there is not much that could be done about it except make the panel aware that it is querying Prometheus or add a method to the Datasource superclass (does one exist?) to be invoked by the panel when computing the auto interval. The former is query specific (at least as currently defined) but it could be pulled up to the panel level, as (in my case, at least) it is mostly determined by Prometheus' sampling/rule eval frequency.

All the resolution does is to then take this adjusted interval and multiply it to get the step parameter (sampling frequency). It is also involved in the 11k limit computation, in that it takes into account that you're only going to e.g. request 1 point in 10, so the limit then essentially becomes 110k.

To take a specific example: my Prometheus instance scrapes targets every 5 seconds and evaluates rules every 10 seconds. In conclusion (assuming Prometheus keeps reasonably close to these intervals) it makes no sense to query for data that is higher resolution than 10 seconds (it actually randomly delays graph data by anywhere up to 10 seconds, but there is no way of asking Prometheus for the raw data -- there is actually, but it's not documented and won't work for expressions). Anyway, I always set a 10s "Min step" on all my queries because I want to avoid wholly unnecessary load on Prometheus. This could be pulled up to the panel level although I can imagine cases where you would want different values per query (e.g. if you were displaying both raw metrics at 5 second original resolution and rules at 10 second original resolution).

If I have a query like `rate(request_count[$__interval])`, then that interval will essentially only be defined by the panel auto interval, 10 second min value and 11k sample limit. I will basically be lower bounded by both my 10s configured minimum and range/11k (mostly the former).

Let's assume an auto interval (panel computed) of 2s and a configured resolution of 1/1. The Prometheus query (produced by this PR) is
`/api/v1/query_range?query=rate(request_count[10s])&start=X&end=Y&step=10`
With a configured resolution of 1/10 it becomes
`/api/v1/query_range?query=rate(request_count[10s])&start=X&end=Y&step=100`
I.e. the interval stays the same, the step is 10x larger.

Suppose X and Y are 100k seconds apart and the panel computed auto interval is 2s. The exact same queries as above will be generated.

Suppose now X and Y are 1M seconds apart and somehow the panel still comes up with a 2s auto interval. With a 1/1 resolution you'll get
`/api/v1/query_range?query=rate(request_count[91s])&start=X&end=Y&step=91`
Note that both the Prometheus query and step parameter have been bumped up to 1M / 11k. With a 1/10 resolution the query becomes
`/api/v1/query_range?query=rate(request_count[10s])&start=X&end=Y&step=100`
Same as before, still under the 11k limit.

And finally, if you go to 10M seconds and 2s auto interval:
`/api/v1/query_range?query=rate(request_count[910s])&start=X&end=Y&step=910`
and
`/api/v1/query_range?query=rate(request_count[91s])&start=X&end=Y&step=910`

> > And alerts I don't know about, but if I were to guess, all they would have in common with this logic is the 11k sample limit.
>
> The code that issues Prometheus queries in the Go backend should behave the same as the frontend code. So if prometheus data source code modifies the build in $__interval and $__interval_ms variables this needs to be done in the backend as well.

I quickly scanned the Grafana alerting documentation and it states "The query used in an alert rule cannot contain any template variables." So does this really apply then?

**Comment by alin-amana** [comment]
To look at it otherwise, if you wanted to display `$__interval` in the title or elsewhere then you would have a number of options:
1. The radical option: pull up both resolution and min interval to the panel level (although not sure how that works with other datasources) and take away the freedom to configure them separately per query. You could then expose datasource specific limitations (e.g. the 11k limit) and do the interval computation at panel level.  I wouldn't like this (particularly the resolution bit), and I'm very resentful. :o)
1. The slightly less radical option: pull min interval up to panel level (again, not sure about other datasources). Optionally also expose datasource specific limitations to the panel and let it compute an interval independent of per-query resolution (pretty much as it is now, but without integrating the 11k limit as nicely). Queries with lower resolution would unnecessarily be limited to 11k divided by resolution instead of 11k, but it wouldn't be all that bad. I'm ambivalent about the result, but it does require quite a bit of rework.
1. Let the queries actually update the panel-level `$__interval` value instead of making a local copy. Hacky, messy, ripe for race conditions.

Or you can stick with the current approach, where you can't use `$__interval` anywhere but in the query. Not that nice, but it exists, doesn't break anything and can be later "upgraded" to any of the other options without any breakage, really.

If I could choose and the choice was entirely free, I would either stick with the current implementation or go with (2) above, mostly because it would make min interval functionality available to other data sources (or, if it already exists, standardize it).

**Comment by torkelo** [comment]
now that step is always a lower limit, hitting the 11K limit should not be possible unless you had a 12K monitor

**Comment by alin-amana** [comment]
Friendly ping.

**Comment by SuperQ** [comment]
> To take a specific example: my Prometheus instance scrapes targets every 5 seconds and evaluates rules every 10 seconds. In conclusion (assuming Prometheus keeps reasonably close to these intervals) it makes no sense to query for data that is higher resolution than 10 seconds

This is not always true.  If you are querying metric without a recording rule, and there are 100 instances, you are effectively getting enough data for a resolution below 10 seconds per step.

**Comment by alin-amana** [comment]
Fair enough. All my dashboards thus far only display job-level stats, whether from recording rules or computed on the fly from scraped metrics. In this particular situation, graphing on-the-fly computed data with higher than 5 seconds resolution merely gives me annoying repeated patterns (e.g. a spike for instance 0, followed by a trough for instance 1) every 5 seconds. I actually prefer the smoothing effect that limiting the resolution gives.

So, to be more precise, I prefer to limit maximum resolution to 10 seconds because it's more data, less noise, that way.

**Comment by alin-amana** [comment]
With apologies for being a PITA, I'd really like to see this merged. The way I see it (I might be wrong) is there are 2 reasons why this is on hold:

1. It would be nice to have the `$__interval` computation and at least some of the underlying components (resolution, min interval) consistent across data sources and surfaced all the way to the panel (so that e.g. it may be displayed in the panel title).
1. Have any changes to the way interval is being computed in panels reflected in the backend.

That's pretty much it.

As far as I can tell (2) is not an an issue, because (according to documentation) you can't use template variables for alerting. That includes `$__interval`.

As for (1), while I agree it would be nice to have some common logic across data sources as to computing and exposing `$__interval`, that would essentially be a new feature, not having to do with this particular PR. Merging this PR would not impede that work in any way. The actual interval used for Prometheus queries already differs from `$__interval` and varies per query. All this PR does is fix that computation.

Am I missing anything?

**Comment by torkelo** [comment]
> As far as I can tell (2) is not an an issue, because (according to documentation) you can't use template variables for alerting. That includes $__interval.

These built in variables can be used in alerting queries.  

Sorry for taking the time to review this further. We are just swamped with so many issues & PRs and at the same time trying to work on strategically important issues & changes. 

Grafana tries to be backward compatible as much as possible so changes like this that we might later want to change is a bit tricky to know how to handle. But as you say step and interval is not aligned right now either, but looking closer it looks like this PR does not fix that either, the adjustInterval function does not return an adjusted interval based on the resolution. Should it not do that?

**Comment by alin-amana** [comment]
> These built in variables can be used in alerting queries.

Did not know that. Like I said, I never used Grafana alerts. Will look into it, in this case, and get back with a better answer than "not applicable".

> Sorry for taking the time to review this further. We are just swamped with so many issues & PRs and at the same time trying to work on strategically important issues & changes.

Fully understand. Hence the apology for being a PITA.

> Grafana tries to be backward compatible as much as possible so changes like this that we might later want to change is a bit tricky to know how to handle. But as you say step and interval is not aligned right now either, but looking closer it looks like this PR does not fix that either, the adjustInterval function does not return an adjusted interval based on the resolution. Should it not do that?

No. As I see it, the interval should entirely ignore the resolution (apart from calculations regarding the 11k data point limit). The interval, as computed by the panel, is essentially the range divided by the screen resolution and that's how it should stay (modulo bumping it up to a min interval, when that is configured). The resolution, as documented, is a way of requesting only a sub-sample of points (e.g. 1/10). So you first compute the interval as if there was no resolution, then, if a resolution is configured, only request one in X points, each point covering `$__interval` seconds.

If you approached it the other way (auto interval of T seconds, resolution of 1/10, set `$__interval` to T*10) and you assume the query actually uses `$__interval` for aggregation then you're still essentially forcing Prometheus to aggregate all data, only at a lower resolution. That's likely better than full resolution in terms of load on Prometheus, but not as good as asking for data points aggregated over T seconds 10*T seconds apart.

If you actually want, say no more than 100 data points across your graph, then you can define an interval template variable with a step count of 100 (and a reasonable min interval), then set both the min step of your query and the interval in your query to the template variable. Then for e.g. a 24 hour range you'll get 100 data points ~15 minutes apart, each covering about ~15 minutes of time. (This is where my other PR to add support for multiple interval variables comes in.) :o)

[For context, I'm thinking here of a counter with a `rate()` or `increase()` applied to it and trying to have each original data point only included in exactly one graph data point.]

I guess the tests illustrate quite well my expectations of how this should work.

**Comment by torkelo** [comment]
I thought the whole point of $__interval and $__interval_ms was to be in sync with step so that you can do accurate queries (for max/min/sums over range vectors). But maybe I have misunderstood something.  

I thought the resolution optimization (and increase in step) was mostly to reduce the number of internal instant queries Prometheus has to do, not a way to skip data completely.

**Comment by alin-amana** [comment]
You can look at it either way, I guess.

When I implemented this I looked at the online documentation and tried to keep things as close to that as possible. In particular, the documentation for $__interval states

> Approximate Calculation: `(from - to) / resolution`

with `resolution` here referring to screen resolution. I assumed someone might rely on that, so (modulo min interval and 11k data point limit) I tried to keep as close to the documentation, i.e. not go `(from - to) * intervalFactor / resolution` (where `intervalFactor` is the Resolution field of the Prometheus query editor). In the vast majority of cases this ends up not changing the value of `$__interval` as passed down by the panel. And it has the added benefit that queries with lower resolution (let's assume 1/10) are much faster when `step == 10 * $__interval`. E.g. a couple of my heavier queries are 4-5x faster when `step == 300` and `$__interval == 30` than when `step == 300` and `$__interval == 300` simply because the instant queries have to work with less data. The queries will end up essentially sampling 1/10 points, but you could argue that's what you get from 1/10 resolution.

On the other hand, you also have a valid point: you will reduce the load on Prometheus by asking for fewer points even if you're not sampling 1/10 data points but instead aggregating over 10x the auto interval. You are then deviating more from the documentation, but you're not sampling (i.e. dropping) data anymore (assuming your query attempts to do that in the first place).

To give you an idea of how fast or slow each approach is, here are some time measurements from a couple of heavy queries I'm using. They both end up doing `rate(foo[$__interval])` with some sums, joins and quantiles on top over a range of 24 hours.

auto interval | intervalFactor | Approach | Outcome | Query 1 time | Query 2 time
---|---|---|---|---|---
30s|1/1|any|`step=30s $__interval=30s` | 8 s | 2 s
30s|1/10|yours|`step=300s $__interval=300s` | 3 s | 0.7 s
30s|1/10|mine|`step=300s $__interval=30s` | 1.5 s | 0.5 s

As you can see, in terms of query time and Prometheus load, there is benefit in lowering the resolution (your approach, ~65% reduction) and even more benefit in sampling (my approach, an extra 30-50% reduction).

If this was my code and I didn't have to care about maintaining a tight API, I would create 2 variables `$__interval` and `$__step` (or something more reasonable), one giving me the interval as computed by the panel, strictly based on pixel resolution (in my examples above, 30s); and the other giving me the `intervalFactor` adjusted interval (or the value sent to Prometheus as the step parameter (in the examples above, 300s). This would leave the option of whether to (a) lower resolution and maintain coverage; or (b) lower resolution and sample; with their respective trade-offs in terms of time and load to whoever is implementing the dashboard.

Since neither I nor you are in this position, I guess all that remains is that you and/or other owners make a decision on whether you want Prometheus query editor's `intervalFactor` to:
1. Lower resolution on queries, with `$__interval` reflecting that lowered resolution, allowing for complete coverage but only improving query latency and Prometheus load partially.
2. Lower resolution on queries, with `$__interval` staying the same, resulting in sampling (i.e. skipping data) and maximum improvement on query latency and Prometheus load.
3. Give users the choice and have 2 separate `$__interval` variables, one pretty much the same as computed by the panel (modulo min interval and 11k limit) and one adjusted to match the `step` query parameter.

Whichever you decide upon, I will attempt to dutifully make the changes. :o)

**Comment by SuperQ** [comment]
The target use case we started with for the `$interval` was brought up by @roman-vynar for dealing with wide zoom ranges in graphs.  We wanted the automatic-interval to be the time distance between two points on the graph.  Essentially allowing for each point on the graph to be an `increase()` function between the two points.  Speed of rendering in some cases is a bonus, but accuracy of the graph is more important.

For example, given a metric with 15s scrape interval, and a 7 day graph displaying 70 data points.  Each data point would be 288s apart.  In this case `$__interval` should be 288s.  This may be a lot of samples to send to the `rate()` function, but it's the intended use case.

**Comment by alin-amana** [comment]
Fair enough, I can live with that. Just figured keeping close to the Grafana documentation would increase my chances of having this PR accepted. Apparently not. :o)

For the record, sampling could be approximated by defining an interval template variable with high step count (say 1000) and using that in your query and reducing the resolution (or `intervalFactor`) in the query editor to 1/10. And with this in sight, I'm going to shamelessly plug my other PR, adding support multiple interval template variables, #9216.

**Comment by SuperQ** [comment]
If the Grafana documentation is wrong or confusing, we should probably fix it. 😄

**Comment by alin-amana** [comment]
Well, it's not so much that it's wrong. Just that the resolution (or `intervalFactor`) parameter is a Prometheus query specific thing and so the Grafana documentation doesn't mention it when explaining how `$__interval` is calculated. It does say "approximate calculation", though.

If, on the other hand, you look at the Prometheus query editor documentation, it does say that resolution controls the step option, without mentioning any downsampling, so it would seem I was wrong to assume that it should result in downsampling the data.

So if @torkelo and @bergquist don't want the more complex option of two separate interval variables (including and ignoring resolution), I should probably go ahead and just multiply by resolution.

Which brings up a minor question: should e.g. a 1/10 resolution (i.e. `intervalFactor=10`) result in a step that's a multiple of 10 or simply whatever `range / pixel_resolution * intervalFactor` yields?

**Comment by alin-amana** [comment]
Regarding server-side use of `$__interval`, i did create an alert in one of my dashboards and:
1. It worked just fine as long as I used an explicit interval, e.g. `rate(foo[5m])`. Good.
2. It explicitly told me template variables were not supported in alerts when I used an interval template variable, e.g. `rate(foo[$period])`. Good again.
3. Did not complain, but failed regardless when I used `$__interval`, e.g. `rate(foo[$__interval])`. Bad.

I used strace to find out what query Grafana was sending and the query included the literal `$__interval`, not replaced by an actual interval of any kind.

In conclusion I would argue that my change doesn't make the usage of `$__interval` more inconsistent in frontend vs. backend, as the backend is just as borked as before. :o(

**Comment by alin-amana** [comment]
I've updated the code so that `$__interval` is now always equal to the step parameter in the Prometheus query. It continues to take into account the 11k data point limit, but ignoring that It is now essentially computed as:
```
max(autoInterval * intervalFactor, minInterval)
```
where:
* `autoInterval` is the interval computed by the panel (i.e. `timeRange / resolution`);
* `intervalFactor` is the *Resolution* field of the Prometheus query editor; and
* `minInterval` is the *Min step* field of the Prometheus query editor.

Some examples of inputs and outputs:

range | autoInterval | minInterval | intervalFactor | `$__interval` | step query param
---|---|---|---|---|---
1h | 2s | - | - | 2s | 2
1h | 2s | 1s | - | 2s | 2
1h | 2s | 5s | - | 5s | 5
1h | 2s | - | 10 | 20 | 20
1h | 2s | 10s | 10 | 20s | 20
1h | 2s | 50s | 10 | 50s | 50
1w | 2s | - | - | 55s | 55
1w | 2s | - | 10 | 60s | 60

**Comment by alin-amana** [comment]
@torkelo, any time to take a quick look and (possibly) merge?

As described [two comments up](https://github.com/grafana/grafana/pull/9226#issuecomment-332481732) backend use of `$__interval` does not work with Prometheus with or without this change, so tying it to backend changes doesn't make much sense in my mind.

As for pulling resolution and/or interval up to the panel level: that might be a nice feature to have standardized, but this change is simply a bugfix and so does not affect whether or how that feature would be implemented in any way.

**Comment by torkelo** [comment]
Sorry, we will try find time to review & refactor this.

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (bergquist) [inline public/app/plugins/datasource/prometheus/datasource.ts]: The code within this loop should be extracted into a new function.

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (bergquist) [inline public/app/plugins/datasource/prometheus/datasource.ts]: Please use https://github.com/grafana/grafana/blob/master/public/app/core/utils/kbn.js#L204 instead of `this.intervalSeconds`.

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (bergquist) [inline public/app/plugins/datasource/prometheus/datasource.ts]: just return directly.

**Review by bergquist — CHANGES_REQUESTED** [review]

**Comment by bergquist** [comment]
Quite a read :)

Lets ignore the discussion about moving resolution/interval up to the panel level for this PR.
I think it fair to also ignore support for interval variables in alerting as well.

Some minor remarks before we merge this.

**Comment by alin-amana** [comment]
Apologies for the sheer quantity of comments. And thanks for the review comments.

All done, PTAL.

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (alin-amana) [reply public/app/plugins/datasource/prometheus/datasource.ts]: Done, function named createQuery().

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (alin-amana) [reply public/app/plugins/datasource/prometheus/datasource.ts]: Done.

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (alin-amana) [reply public/app/plugins/datasource/prometheus/datasource.ts]: Good point, done.

**Review by alin-amana — COMMENTED** [review]
All done, PTAL.

**Comment by bergquist** [comment]
@alin-amana I think this will be very useful. Thank you for taking so much time working on this PR!

**Review by bergquist — APPROVED** [review]

**Comment by alin-amana** [comment]
Sure, np. Thanks everyone for the review and discussion, they turned out to be quite useful.

I might come back and look into the backend bits and/or pulling up the resolution to panel level, but I'm not promising anything. I'll be going on vacation soon and have quite a few loose ends to tie up.

**Pull request merged**
